apsuite/commissioning_scripts/inj_traj_fitting.py

```diff
@@ -0,0 +1,230 @@
+        self.twiss, *_ = pyaccel.optics.calc_twiss(self.model)
+
+    def calc_traj(self, x0, xl0=0, y0=0, yl0=0, delta=0, size=160):
+        """."""
+        rin = np.array([x0, xl0, y0, yl0, delta, 0])
+        rout, *_ = pyaccel.tracking.linepass(
+            self.model, rin, self.bpm_idx[:size])
+        return rout[0, :], rout[2, :]
+
+    def calc_residue(self, vec, tx_meas, ty_meas):
+        """."""
+        tx_mod, ty_mod = self.calc_traj(*vec, size=tx_meas.size)
+        nanidcs = np.logical_or(np.isnan(tx_mod), np.isnan(tx_mod))
+        tx_mod[nanidcs] = 10.0
+        ty_mod[nanidcs] = 10.0
+        return np.hstack([tx_mod-tx_meas, ty_mod-ty_meas])
+
+    def calc_chisqr(self, residue):
+        """."""
+        return np.sum(residue*residue)/residue.size
+
+    def calc_jacobian(self, vec, size=160):
+        """."""
+        mat = np.zeros((2*size, 5))
+        zer = np.zeros(size)
+        dlt = 1e-5
+        for i in range(vec.size):
+            dvec = np.array(vec)
+            dvec[i] += dlt/2
+            res_pos = self.calc_residue(dvec, zer, zer)
+            dvec[i] -= dlt
+            res_neg = self.calc_residue(dvec, zer, zer)
+            mat[:, i] = (res_pos - res_neg)/dlt
+        return mat
+
+    def calc_init_vals(self, trajx, trajy):
+        """."""
+        etax_ave = np.mean(self.twiss.etax[self.bpm_idx])
+        x_ini, y_ini, xl_ini, yl_ini = trajx[0], trajy[0], 0, 0
+        de_ini = np.mean(trajx) / etax_ave
+        return np.array([x_ini, xl_ini, y_ini, yl_ini, de_ini])
+
+    def do_fitting(self, trajx, trajy, vec0=None, max_iter=5, tol=1e-6):
+        """."""
+        vec0 = vec0 if vec0 is not None else self.calc_init_vals(trajx, trajy)
+        res0 = self.calc_residue(vec0, trajx, trajy)
+        chi0 = self.calc_chisqr(res0)
+        vecs = [vec0, ]
+
+        vec, res = vec0.copy(), res0.copy()
+        factor = 1
+        for _ in range(max_iter):
+            mat = self.calc_jacobian(vec0, size=trajx.size)
+            u_mat, s_mat, vh_mat = np.linalg.svd(mat, full_matrices=False)
+            imat = vh_mat.T @ np.diag(1/s_mat) @ u_mat.T
+
+            dpos = imat @ res
+            vec -= dpos * factor
+            res = self.calc_residue(vec, trajx, trajy)
+            chi = self.calc_chisqr(res)
+            if chi >= chi0:
+                vec = vec0.copy()
+                res = res0.copy()
+                factor = max(factor/2, 1e-2)
+            else:
+                vec0 = vec.copy()
+                res0 = res.copy()
+                vecs.append(vec0)
+                chi0 = chi
+                factor = min(factor*2, 1)
+            if chi0 < tol:
+                break
+        return vecs
+
+    def get_traj_from_sofb(self):
+        """."""
+        trajx = self.devices['sofb'].trajx.copy()
+        trajy = self.devices['sofb'].trajy.copy()
+        summ = self.devices['sofb'].sum.copy()
+
+        ini = np.mean(summ[:3])
```

Comment:
maybe this could be an input with default value of 3 or an attribute of `Params` just like `count_rel_thres`, since these two parameters seems to be closely related.

Reply:
Yeah, that's a good idea. I'll do it!

Reply:
done!